validation(get_branch_version_from_list): Check that URL is valid and…

Orenef11 opened the pull request:
* Refactor function `get_branch_version_from_list` & `get_branch_version_from_repo` to extract version from any URL
 * Added validation for all URLs in `sct_config`
 * Fix files yaml:
   * `complex_test_case_with_version.yaml`
   * `multi_region_dc_test_case.yaml`

Trello url: https://trello.com/c/y2WdQpkz/1777-verify-links-in-repo-and-list-files

## PR pre-checks (self review)
<!--- PR should be created as Draft, when CI finished and relevant checkboxes selected, add reviewers and then click on "Ready for review" button.-->
<!--- Put an `x` in all the boxes that apply or create PR and then click on all relevant checkboxes: -->
- [x] I followed [KISS principle](https://en.wikipedia.org/wiki/KISS_principle) and [best practices](https://docs.google.com/document/d/1jihgOKb5iGRlD8_HQ92O0JbLk1kASUoZT23i_MXFSKI)
- [x] I gave variables/functions meaningful self-explanatory names
- [x] I didn't leave commented-out/debugging code
- [x] I didn't copy-paste code
- [x] I added the relevant `backport` labels
- [x] New configuration option are added and documented (in `sdcm/sct_config.py`)
- [x] I have added tests to cover my changes (Infrastructure only - under `unit-test/` folder)
- [x] All new and existing unit tests passed (CI)
- [x] I have updated the Readme/doc folder accordingly (if needed)

fruch commented:
Also check the pre-commit run in Jenkins

https://jenkins.scylladb.com/job/sct-github-PRs-scan/job/scylla-cluster-tests/job/PR-1996/4/execution/node/25/log/

It's not happy with your commit message

Orenef11 commented:
Retry DeepCode

fruch commented on sdcm/db_stats.py:
```diff
@@ -559,51 +559,58 @@ def calculate_stats_total(self):
         self._stats['results']['stats_total'] = total_stats
 
     def update_test_details(self, errors=None, coredumps=None, scylla_conf=False, dict_specific_tested_stats=None):
-        if self.create_stats:
```
Why are you touching me this code ?

roydahan commented:
Can someone explain me what the commits content of this PR?
How is PR from Oren has 1 commit from Alex?

Orenef11 commented:
@roydahan 
My mistake!
I was `amend` to Alex's commit

Orenef11 commented:
@fgelcer 
All failures because  some of `yaml` files contain invalid url (`http://repositories.scylladb.com`)

fruch commented on requirements-python.txt:
```diff
@@ -20,8 +20,6 @@ PTable==0.9.2
 pycryptodome==3.8.1
 pyzmq==18.0.1
 tenacity==5.0.4
-# this package isn't in pypi
```
why are you removing this, please leave this package in

fruch replied:
if you want to remove this clone from private branch, please code all the code into our repo.
why rewriting it ?

fruch replied:
you can opne a trello task to handle this repo, i.e. by uploading to pypi
or by copying it into out code base

fruch commented on sdcm/sct_config.py:
```diff
@@ -1044,6 +1045,15 @@ def __init__(self):
         if self.get('instance_provision') == 'mixed':
             LOGGER.warning('Selected instance_provision type "MIXED" is not supported!')
 
+        urls = [self.get('new_scylla_repo', None), self.get('scylla_repo_m', None),
```
please move this check into `verify_configuration` function. and into the constructor

fruch replied:
a new flag ? why ?

fruch commented on sdcm/utils/xml_utils.py:
```diff
@@ -0,0 +1,54 @@
+import re
```
why are you rebuilding the repoParser module ? how it's related to what this task is about ?

fruch commented on sdcm/utils/url_validate.py:
```diff
@@ -0,0 +1,50 @@
+
+import requests
+
+SUPPORTED_WEBSITES = ['s3.amazonaws.com', 'scratch.scylladb.com']
```
please don't limit the places we could get repo from like that.

the idea is to check we could use that repo, not to limit it to specific hosts

Orenef11 replied:
Yes, you're right, but to validate the URL I need to know the name of hosts.
For new URL we need to update this list.
What do you thinking?

fgelcer replied:
you can get if from the string you get... but, why is it important to you to know the name of hosts? what if a developer created a temporary repo to run tests for his tests?

fgelcer replied:
i don't understand what it means... the .repo or .list files could point anywhere and still be valid.. they are supposed to be repositories, and don't need to be hosted by scylla.com

fruch commented on sdcm/utils/url_validate.py:
```diff
@@ -0,0 +1,50 @@
+    Check the URL of `scylla.list` is valid and verify it's a file.
+    After that check, the file contains a valid scheme and package URL.
+    """
+    supported_content_types, package_urls = ['application/octet-stream', 'binary/octet-stream'], []
```
this a very http server specific, I don't think we should care much about those headers...

Orenef11 replied:
To verify that the URL can be downloaded, it's must contain one of these headers.
Yes, it's very specific but we want to verify each step.
In the future, if we will get failures I'll change this line to less **strict**.

fruch commented on sdcm/utils/version_utils.py:
```diff
@@ -2,66 +2,45 @@
-    except ValueError:
-        raise ValueError("url isn't a correct deb list or yum repo\n\turl:[{}]".format(url))
+    get_version_method = get_branch_version_from_list
+    if url.endswith(('.list', '.repo')):
```
why adding this limit of file extension ?

Orenef11 replied:
Because the first URL must contain one of the following files.
Also, we have different logic for each file.

fgelcer replied:
i don't see any logic for .list files

fgelcer commented on sdcm/utils/url_validate.py:
```diff
@@ -0,0 +1,50 @@
+    supported_content_types, package_urls = ['application/octet-stream', 'binary/octet-stream'], []
+
+    if not url.startswith('http'):
+        url = f"{'http://'}{url}"
```
what if someone wants to use FTP, NFS or even HTTPS??

fgelcer replied:
is this URL the one inside the .repo or .list files?

Orenef11 replied:
@fgelcer 
Do you have the `FTP` or `NFS` URL that I can check?

fgelcer replied:
nope, but we can try to add them for future tests

Orenef11 commented:
Some URLs contain invalid host anme `http://repositories.scylladb.com`

fgelcer commented:
> Some URLs contain invalid host anme `http://repositories.scylladb.com`

by looking at it, one cannot know if it is valid or not, right?
so, you should check the URL either it is within the expected names or not, but with some specific content inside of it, right?

bentsi commented on sdcm/utils/url_validate.py:
```diff
@@ -0,0 +1,41 @@
+}
+
+
+def verify_url_is_valid(url):
```
name of the function and package is misleading
URL is a too broad term, this function doesn't check general URL but URLs in the repo files...

Orenef11 replied:
Done, replace the method to `verify_scylla_file_version_is_valid`

bentsi commented on sdcm/utils/url_validate.py:
```diff
@@ -0,0 +1,41 @@
+    (re.compile(r'deb\s+(?P<url>http.*?)\s(?P<version_code_name>.*?)\s(?P<component>.*?)$'),
+     "{url}/dists/{version_code_name}/{component}/binary-amd64/Packages.gz"),
+]
+URL_ENV_VARIABLES = {
```
what about `$basearch=noarch`?

bentsi replied:
example: http://downloads.scylladb.com/rpm/unstable/centos/branch-4.0/2020-04-05T12:42:44Z/scylla/noarch/

Orenef11 replied:
The variable `basearch` can contain more options?

bentsi commented on sdcm/utils/url_validate.py:
```diff
@@ -0,0 +1,41 @@
+            raise ValueError(error_msg)
+    if not package_urls:
+        raise ValueError(error_msg)
+    return package_urls[0]
```
what is the purpose of this return?

Orenef11 replied:
I tried to supported in old logic, Ill change the logic of  `get_branch_version_from_list` and `get_branch_version_from_repo`

Orenef11 commented on sdcm/sct_config.py:
```diff
@@ -1131,7 +1132,7 @@ def check_required_files(self):
                                              f"'{profile_path}' that does not exists under data_dir/")
 
-    def verify_configuration(self):  # pylint: disable=too-many-statements
+    def verify_configuration(self, verify_scylla_url_version=True):  # pylint: disable=too-many-statements
```
New variable without used, the goal of this variable to allow the use of an incorrect Scylla URL version.

fruch replied:
it can be removed, there's no case we are calling without this check.

Orenef11 commented on sdcm/sct_config.py:
```diff
@@ -1232,6 +1233,16 @@ def _check_backend_defaults(backend, required_params):
 
+        if not verify_scylla_url_version:
+            return
+        urls = [self.get('new_scylla_repo'), self.get('scylla_repo_m'), self.get('scylla_repo_loader'),
```
I'm a little confused ...
I added validation for each part of the UL (host address, scheme, response status, etc.) and was told it was too strict.
What kind of validations do I need to add?

fruch replied:
it just meant you can remove the `LOGGER.info(f"Checking if the url '{url}' is valid")`

Orenef11 commented on unit_tests/test_version_utils.py:
```diff
@@ -1,16 +1,19 @@
-RPM_URL = 'https://s3.amazonaws.com/downloads.scylladb.com/enterprise/rpm/unstable/centos/'\
-          '9f724fedb93b4734fcfaec1156806921ff46e956-2bdfa9f7ef592edaf15e028faf3b7f695f39ebc1'\
-          '-525a0255f73d454f8f97f32b8bdd71c8dec35d3d-a6b2b2355c666b1893f702a587287da978aeec22/71/scylla.repo'
+ROOT_A3_DOWNLAND_PAGE = 'https://s3.amazonaws.com/downloads.scylladb.com'
```
Done

Orenef11 commented on unit_tests/test_version_utils.py:
```diff
@@ -1,16 +1,19 @@
-          '9f724fedb93b4734fcfaec1156806921ff46e956-2bdfa9f7ef592edaf15e028faf3b7f695f39ebc1'\
-          '-525a0255f73d454f8f97f32b8bdd71c8dec35d3d-a6b2b2355c666b1893f702a587287da978aeec22/71/scylla.repo'
+ROOT_A3_DOWNLAND_PAGE = 'https://s3.amazonaws.com/downloads.scylladb.com'
+DEB_URL = '{}/enterprise/deb/unstable/stretch/9f724fedb93b4734fcfaec1156806921ff46e956-2bdfa9f7ef592edaf15e028faf3b7f' \
```
Done

fruch commented on sdcm/utils/threads_utils.py:
```diff
@@ -0,0 +1,16 @@
+from repodataParser.RepoParser import Parser
+
+
+class ThreadWithReturnValue(Thread):
```
we all ready have 
https://github.com/scylladb/scylla-cluster-tests/blob/00302b48e51729b8ffafb506bd34f6efafd755ef/sdcm/utils/common.py#L277

that can do the same thing, no point of re-implementing it

fruch commented on sdcm/utils/url_utils.py:
```diff
@@ -0,0 +1,89 @@
+
+from sdcm.utils.threads_utils import ThreadWithReturnValue
+
+SCYLLA_URL_REGEX = re.compile('.*(?P<url>http.*scylladb.com.*).*')
```
again, why are we limiting just to scylladb.com ?

this should work for any reachable urls....
event if I host my own repo on my laptop, and give the public address of it.

fruch replied:
but you are always validating the URL, and we won't be able to run such repos, pointing to something which is not scylladb.com

bentsi replied:
I still don't understand why do we need to distinguish between Scylla and non-Scylla URLs

Orenef11 commented on unit_tests/test_config.py:
```diff
@@ -75,6 +75,7 @@ def test_06a_docker_latest_no_loader():
         conf.verify_configuration()
 
     @staticmethod
+    @unittest.skip
```
@fruch 
The test is failed and not because of my changes.
Can you check it?

fruch commented on unit_tests/test_version_utils.py:
```diff
@@ -19,20 +22,23 @@
 class TestVersionUtils(unittest.TestCase):
     def test_01_get_branch_version_from_list(self):
-        self.assertEqual(get_branch_version_from_list(DEB_URL), '2019.1.1')
+        _ = [self.assertEqual(url_version, '2019.1.1') for url_version in get_branch_version_from_list(DEB_URL)]
```
if the list comes back empty, this test would pass...

Orenef11 replied:
Because the method `get_branch_version` checking if the file with extension  `list` or `repo`

fruch commented on sdcm/utils/url_utils.py:
```diff
@@ -0,0 +1,83 @@
+    for url in urls:
+        if not SCYLLA_URL_REGEX.match(url):
+            raise ValueError(f"The following URL not supported '{url}'")
+    result = [res.result for res in ParallelObject(objects=urls, timeout=SCYLLA_URL_RESPONSE_TIMEOUT).run(
```
please take the parallel logic out of this function, and use it only where it's needed.

don't change `get_branch_version` cause of this, it make the code calling it not that clear, see the unittest that were changed cause of it.

Orenef11 replied:
I've create function for parallel logic and the function `get_branch_version` returns only one value

Orenef11 commented on unit_tests/test_version_utils.py:
```diff
@@ -19,20 +22,23 @@
-        self.assertRaisesRegex(ValueError, "url isn't a correct", get_branch_version_from_list, BROKEN_URL)
-        self.assertRaisesRegex(ValueError, "url isn't a correct", get_branch_version_from_repo, BROKEN_URL)
+        err_msg = "The following URL not supported"
+        self.assertRaisesRegex(ValueError, "The following url", get_branch_version, BROKEN_URL)
```
Because the method `get_branch_version` checking if the file with extension  `list` or `repo`

Orenef11 replied:
Because the method `get_branch_version` checking if the file with extension  `list` or `repo`

Orenef11 commented on unit_tests/test_version_utils.py:
```diff
@@ -19,20 +22,23 @@
     def test_03_get_branch_version(self):
-        self.assertEqual(get_branch_version(RPM_URL), '2019.1.1')
-        self.assertEqual(get_branch_version(DEB_URL), '2019.1.1')
+        _ = [self.assertEqual(url_version, '2019.1.1')
```
Yes I know...
Url file can return multiple values

bentsi commented:
@Orenef11 please address comments and go over the pre-checks again, you checked all of them but without doing the actual self-review and following what is asked

fruch commented on sdcm/utils/url_utils.py:
```diff
@@ -0,0 +1,87 @@
+        file_extension = _url.rsplit('/', maxsplit=1)[1]
+        if file_extension.endswith(('xml', 'xml.gz',)):
+            is_xml_file = True
+        if not file_extension.endswith(SUPPORTED_FILE_EXTENSIONS):
```
remove this condition, we don't care the file extensions, we just read them.
and we shouldn't care about if it's a scylla domain or not

fruch commented on sdcm/utils/url_utils.py:
```diff
@@ -0,0 +1,87 @@
+                raise TypeError(f"The file extension '{file_extension}' not supported")
+            else:
+                return []
+        response = urlopen(url=_url)
```
why are you using urlopen and not requests ?

Orenef11 replied:
@fruch 
For our case it is much faster than `requests.get`

fruch commented on sdcm/utils/url_utils.py:
```diff
@@ -0,0 +1,87 @@
+
+        file_extension = _url.rsplit('/', maxsplit=1)[1]
+        if file_extension.endswith(('xml', 'xml.gz',)):
+            is_xml_file = True
```
why do we have this `is_xml_file` property, I don't see it being used in the code

fruch commented on sdcm/utils/common.py:
```diff
@@ -310,6 +310,8 @@ def __init__(self, objects: Iterable, timeout: int = 6,  # pylint: disable=redef
         :param num_workers: num of parallel threads, defaults to None
         :param disable_logging: disable logging for running func, defaults to False
         """
+        if isinstance(objects, Iterable) and not isinstance(objects, list):
```
do this on the caller, and just fix the typing to the correct one
I think it should be
https://docs.python.org/3/library/typing.html#typing.SupportsIndex

Orenef11 replied:
I removed this change

bentsi commented:
@Orenef11 is it still draft?

Orenef11 commented on sdcm/utils/version_utils.py:
```diff
@@ -1,68 +1,87 @@
+        raise ValueError("The following url '{}' not supported".format(url.rsplit('/', maxsplit=1)[1]))
+
+
+def _run_method_in_parallel(objects, timeout, func, unpack_objects=False):
```
removed

Orenef11 commented on sdcm/utils/version_utils.py:
```diff
@@ -1,68 +1,87 @@
+
+
+def _run_method_in_parallel(objects, timeout, func, unpack_objects=False):
+    result = next((res.result[0] if isinstance(res.result[0], list) else res.result for res in ParallelObject(
```
Done

Orenef11 commented on unit_tests/test_config.py:
```diff
@@ -299,6 +300,7 @@ def test_15_new_scylla_repo(self):
         os.environ['SCT_SCYLLA_REPO'] = centos_repo
         os.environ['SCT_NEW_SCYLLA_REPO'] = centos_repo
         os.environ['SCT_USER_PREFIX'] = 'testing'
+        os.environ['SCT_N_DB_NODES'] = '3'
```
@fruch 
I got an error because this variable contains one value...

fruch replied:
why did you added it ?

Orenef11 replied:
The test failed because the `TypeError`:
```
TypeError: int() argument must be a string, a bytes-like object or a number, not 'NoneType'
```

fruch commented on sdcm/utils/version_utils.py:
```diff
@@ -1,68 +1,205 @@
+    return scylla_file_version
+
+
+def extract_all_scylla_urls_from_file(urls, url_regex_details=None):
```
`url_regex_details` param is not used, please remove

fruch commented on sdcm/utils/version_utils.py:
```diff
@@ -1,68 +1,205 @@
+]
+
+SCYLLA_URL_REGEX = re.compile(".*(?P<url>http.*scylladb.com.*).*")
+UrlDetails = namedtuple("UrlDetails", ["is_xml_file", "url"])
```
please remove `UrlDetails` since `is_xml_file` isn't being used

fruch commented on sdcm/utils/version_utils.py:
```diff
@@ -73,12 +210,7 @@ def is_enterprise(version):
     return parse_version(version) > parse_version('2000')
 
 
-GEMINI_VERSION_RE = re.compile(r'\s(?P<gemini_version>([\d]+\.[\d]+\.[\d]+)?),')
```
why are you removing this code ?

Orenef11 replied:
Moved to line  `18`

Orenef11 replied:
Moved to top on file (line 18)

fruch commented on unit_tests/test_config.py:
```diff
@@ -75,6 +75,7 @@ def test_06a_docker_latest_no_loader():
         conf.verify_configuration()
 
     @staticmethod
+    @unittest.skip("The value of variable 'SCT_SCYLLA_REPO_LOADER' is incorrect")
```
please let the test fail, so it's visible in the PR that this test fail, 
don't skip tests, only if there's a good reason for it.

fruch commented on sdcm/utils/url_utils.py:
```diff
@@ -0,0 +1,87 @@
+}
+
+UrlDetails = namedtuple("UrlDetails", "is_xml_file urls")
+UrlDetails.__new__.__defaults__ = (None,) * len(UrlDetails._fields)
```
there is better way to have default values for namedtuple, but anyhow namedtuple isn't really needed

Orenef11 commented on sdcm/sct_config.py:
```diff
@@ -1294,7 +1294,13 @@ def _check_backend_defaults(backend, required_params):
                     tags = get_ami_tags(ami_id, region_name)
                     assert 'user_data_format_version' in tags.keys(), \
-                        f"\n\t'user_data_format_version' tag missing from [{ami_id}] on {region_name}\n\texisting tags: {tags}"
+                        f"\n\t'user_data_format_version' tag missing from [{ami_id}] on {region_name}\n\texisting " \
```
I changed this line because I got Pylint error

Orenef11 commented on unit_tests/test_version_utils.py:
```diff
@@ -1,16 +1,19 @@
-          '9f724fedb93b4734fcfaec1156806921ff46e956-2bdfa9f7ef592edaf15e028faf3b7f695f39ebc1'\
-          '-525a0255f73d454f8f97f32b8bdd71c8dec35d3d-a6b2b2355c666b1893f702a587287da978aeec22/71/scylla.repo'
+BASE_S3_DOWNLOAD_URL = 'https://s3.amazonaws.com/downloads.scylladb.com'
+DEB_URL = \
```
For some reason, the `pre-commit` or `Pylint` tries to arrange the length of the row.
Therefore, after arranging we get that the first row is not on the same line if the other rows.

fruch commented on sdcm/utils/version_utils.py:
```diff
@@ -4,65 +4,161 @@
+            scylla_patterns=get_all_scylla_url_patterns(url=url), url_regex_details=URL_REGEX_AND_FORMAT_LIST))
+
+    for thread in ParallelObject(objects=list(all_scylla_version_urls), timeout=SCYLLA_URL_RESPONSE_TIMEOUT).run(
+            func=validate_url_scylla_version_file, ignore_exceptions=True):
```
why `ignore_exceptions=True` ?
ending up with only `VERSION_NOT_FOUND_ERROR` as the error is **all** of those url fail the check, isn't helping much

Orenef11 replied:
You're right, that's the purpose of the function.
But, the file can contain lots of links and some of them are not Scylla's.
The primary purpose of the function is to make sure that the version of the version is correct or not.
Unfortunately, I have to ignore errors because I get an error for manger links.
If necessary, I can add special treatment for this manger case.

fruch commented on sdcm/utils/version_utils.py:
```diff
@@ -4,65 +4,161 @@
+VERSION_NOT_FOUND_ERROR = "URL isn't a correct deb list or yum repo"
+
+
+def get_all_scylla_url_patterns(url):
```
this function is getting one url, why it is called `get_all_*` ?

Orenef11 replied:
Although the function receives one URL, it extracts all the patterns that have `baseurl` prefix.
A file can contain multiple patterns (some of them not Scylla URL), so I called the function like that.

fruch commented on sdcm/utils/version_utils.py:
```diff
@@ -4,65 +4,166 @@
+                continue
+
+            version_url = url_format.format(**packages_details.groupdict())
+            is_base_arch_found = BASE_ARCH_KEY in version_url
```
the logic is too complex to grasp what's going on,

try something like that, using the built in template to help:
```python
from string import Template

all_scylla_version_urls = set()
...
# for scylla-manager we never a noarch
arch_list =  ["x86_64", "noarch"] if 'scylla-manager' not in version_url else ["x86_64"]

for arch in ["x86_64", "noarch"]:
     all_scylla_version_urls.add(Template(version_url).substitute(basearch=arch, releasever='7'))
```

also the excessive use of variable for each thing, make this very hard to follow, 
since you need to scroll back and forth to understand what is what.

fruch commented on sdcm/utils/version_utils.py:
```diff
@@ -1,68 +1,152 @@
+            version_url = url_format.format(**packages_details.groupdict())
+            # for scylla-manager we never a noarch
+            basearch_list = ["x86_64", "noarch"] if 'scylla-manager' not in version_url else ["x86_64"]
+            if '$' in version_url:
```
this if/else isn't needed, please remove

Orenef11 commented on sdcm/utils/version_utils.py:
```diff
@@ -1,70 +1,151 @@
 
-    versions_regex = re.compile(r'Version: (.*?)\n', re.DOTALL)
-    major_versions = [m.groups()[0].split('-')[0] for m in versions_regex.finditer(page)]
+def extract_all_urls_from_scylla_patterns(scylla_patterns, url_regex_details):
```
Thanks, fixed

Orenef11 commented on sdcm/sct_config.py:
```diff
@@ -1307,7 +1307,13 @@ def _check_backend_defaults(backend, required_params):
+                        f"\n\t'user_data_format_version' tag missing from [{ami_id}] on {region_name}\n\texisting " \
+                        f"tags: {tags}"
+
+        url_variable_names = ['new_scylla_repo', 'scylla_repo_m', 'scylla_repo_loader', 'scylla_mgmt_repo',
+                              'scylla_mgmt_agent_repo', 'scylla_mgmt_agent_repo']
+        extract_all_scylla_urls_from_file(urls=[self.get(
+            url_variable_name) for url_variable_name in url_variable_names if self.get(url_variable_name)])
```
@bentsi 
Thanks for your comments
1. The function `extract_all_scylla_urls_from_file` returns list of scylla_version and also validate each URL.
For this case, the result is unnecessary.

2.The main goal is not to affect performance, as much as possible, when extracting URLs from the file.
Sometimes a request for a request can take a long time, so there is one function that returns and checks every link found.
I can separate it into some useless functions (because I only used them to create one function).

3. I fixed the logic of those lines

Orenef11 commented:
@bentsi 
I can't respond to your comment, so I'm writing a new comment.
[your comment](https://github.com/scylladb/scylla-cluster-tests/pull/1996#pullrequestreview-415240452)

I fixed some of the notes, changed the function names and also added a new logic for checking the repo files.

The last part I cannot do, this one limits the files which can be.
Another more important thing, I do not know the difference between them, so I can build for each other function ...

fruch commented:
@Orenef11 please fix the unittest

Orenef11 commented:
@fruch @bentsi 
* All comments are resolved.
* The logic has been split to sub-functions.
* There are savings in code lines :-)

Orenef11 commented:
> 1. I am sorry, but I am not able to check this: you didn't follow rule 1 and  mostly rule #4 in the "Coding - best practices". Think of it, how one who never seen repo file/debian repo format can read the code and check its correctness? I am not talking about when it fails.
> 2. naming: when you are saying that the function extracts something, one who reads the code, will understand that it should do some actions and return values. In your case, it doesn't return anything. Your function validates, but not extracts. Find a better name!
> 3. I've already told you about it: from an SW architectural point of view: first of all, solve a problem for one case, then you know how to solve it for N.
>    i.e.
>    Your solution should look something like this:
> 
> ```python
> for repo_url_param in ['new_scylla_repo', 'scylla_repo_m', 'scylla_repo_loader', 'scylla_mgmt_repo', 'scylla_mgmt_agent_repo', 'scylla_mgmt_agent_repo']
>       validate_repo_file_from_url(url=self.get(repo_url_param))
> ```
> 
> That means that `validate_repo_file_from_url` should handle **one** URL. As I asked in my previous comments you should divide the problem into 2 subproblems one for Debian format repository file and second YUM repository format.
> so your `validate_repo_file_from_url` should look something like:
> 
> ```python
> def validate_repo_file_from_url(url):
>     repository_type = get_repo_type(url)
>     if repository_type == "debian":
>         validate_debian_repo()
>     elif repository_type == "yum":
>         validate_yum_repo()
> ```
> 
> and again (this is last time I am explaining it) solve Debian and YUM problems separately (in separate functions)
> In computer science it is called [separation of concerns (SoC)](https://en.wikipedia.org/wiki/Separation_of_concerns) principle

fixed

Orenef11 commented:
> LGTM

@bentsi  What do you think?

fruch commented:
@Orenef11 please fix the conflict

Orenef11 commented:
> @Orenef11 please fix the conflict

Done :-)

fruch commented:
@Orenef11 can you also squash it into one commit ?

roydahan commented:
@fruch you can also squash during merge

fruch commented:
> @fruch you can also squash during merge

I know, but I prefer @Orenef11 would do the editing of his commit message.
and happened to me in some cases that I've forgot to squash and used the rebase and merge.
since in some case we do want to preserve the commits...

Orenef11 commented:
> @fruch you can also squash during merge

> > @fruch you can also squash during merge
> 
> I know, but I prefer @Orenef11 would do the editing of his commit message.
> and happened to me in some cases that I've forgot to squash and used the rebase and merge.
> since in some case we do want to preserve the commits...

NP :-)

fruch commented on sdcm/utils/version_utils.py:
```diff
@@ -1,68 +1,150 @@
+    return urls
+
+
+def get_branch_version_from_debian_repository(urls):
```
why one function is called `get_branch_version_from_debian_repository` and the other `validate_yum_repository` ?
they are both doing the same thing, 
also pick one deb/rpm, repo/list, debian/centos, don't mix them since it's confusing

Orenef11 replied:
Thanks!

fruch commented on sdcm/utils/version_utils.py:
```diff
@@ -1,68 +1,150 @@
+            return ""
+        return max(set(major_versions), key=major_versions.count)
+
+    threads = ParallelObject(objects=urls, timeout=SCYLLA_URL_RESPONSE_TIMEOUT).run(func=get_version)
```
please for the get version, don't use ParallelObject, it might get us the wrong version, since we are not picking it by name.
think if we have 
```
baseurl=https://repositories.scylladb.com/scylla/downloads/scylladb/qa-test/scylla-enterprise/rpm/centos/scylladb-2019.1/$releasever/$basearch/
baseurl=https://repositories.scylladb.com/scylla/downloads/scylladb/qa-test/scylla-enterprise/rpm/centos/scylladb-2019.1/$releasever/noarch/
baseurl=https://copr-be.cloud.fedoraproject.org/results/scylladb/scylla-3rdparty/epel-7-$basearch/
```
if in each URL we'll have packages with versions like `5.7.9` or `3.8`, and scylla is `3.4`, we might endup with getting `5.7.6` since it the max version

Orenef11 replied:
No, the function only finds the Scylla versions.
The last line `baseurl=https://copr-be.cloud.fedoraproject.org/results/scylladb/scylla-3rdparty/epel-7-$basearch/` cannot be matched

fruch replied:
o.k. we'll do with it, the code before that was picking only one url for getting the versions

Orenef11 commented:
@bentsi @fruch  
All comments are resolved :-)
